src/Juvix/Core/Evaluator.hs

```diff
@@ -0,0 +1,135 @@
+{-# LANGUAGE BangPatterns #-}
+{-# OPTIONS_GHC -Wno-unrecognised-pragmas #-}
+
+{-# HLINT ignore "Avoid restricted extensions" #-}
+{-# HLINT ignore "Avoid restricted flags" #-}
+
+module Juvix.Core.Evaluator where
+
+import Control.Exception qualified as Exception
+import Data.HashMap.Strict ((!))
+import GHC.Show
+import Juvix.Core.Extra
+import Juvix.Core.Language
+import Juvix.Core.Language.Info qualified as Info
+import Juvix.Core.Types.InfoTable
+
+newtype EvalError = EvalError String
+
+instance Show EvalError where
+  show :: EvalError -> String
+  show (EvalError msg) = "evaluation error: " ++ msg
+
+instance Exception.Exception EvalError
+
+-- We definitely do *not* want to wrap the evaluator in an exception monad / the
+-- polysemy effects! This would almost double the execution time. Evaluation
+-- errors should not happen for well-typed input (except perhaps division by
+-- zero), so it is reasonable to catch them only at the CLI toplevel and just
+-- exit when they occur.
+
+-- `eval ctx env n` evalues a node `n` whose all free variables point into
+-- `env`. All nodes in `ctx` must be closed. All nodes in `env` must be values.
+-- Invariant for values v: eval ctx env v = v
+eval :: IdentContext -> Env -> Node -> Node
+eval !ctx !env0 = convertRuntimeNodes . eval' env0
+  where
+    evalError :: String -> a
+    evalError msg = Exception.throw (EvalError msg)
+
+    eval' :: Env -> Node -> Node
+    eval' !env !n = case n of
+      Var _ idx -> env !! idx
+      Ident _ sym -> eval' [] (ctx ! sym)
+      Constant {} -> n
+      Axiom {} -> n
+      App _ l r ->
+        -- The semantics for evaluating application (App l r) is:
+        --
+        -- eval env (App l r) =
+        --   case eval env l of
+        --     Closure env' b ->
+        --       let !v = eval env r in eval (v : env') b
+        --
+        -- To do this more efficently for multi-argument functions (without
+        -- creating closures for each intermediate function and matching each
+        -- twice) we gather all application arguments, evaluate them from left
+        -- to right and push the results onto the environment.
```

Comment:
Probably this doesn't make much difference since I changed Builtin and Constructor to BuiltinApp and ConstructorApp.